scipy/special/_basic.pyi

```diff
@@ -0,0 +1,271 @@
+from __future__ import annotations
+
+from typing import (
+    Callable,
+    overload,
+    Literal,
+)
+
+from ._ufuncs import psi
+from numpy import sinc as sinc
+
+from numpy.typing import NDArray
+
+from scipy._lib._util import DecimalNumber, IntNumber
+
+_ComplexNumber = DecimalNumber | complex
+
+_PointsDerivate = tuple[NDArray, NDArray]
+_ZerosOrder = tuple[NDArray, NDArray, NDArray, NDArray]
+
+__all__ = [
+    "ai_zeros",
+    "assoc_laguerre",
+    "bei_zeros",
+    "beip_zeros",
+    "ber_zeros",
+    "bernoulli",
+    "berp_zeros",
+    "bi_zeros",
+    "clpmn",
+    "comb",
+    "digamma",
+    "diric",
+    "erf_zeros",
+    "euler",
+    "factorial",
+    "factorial2",
+    "factorialk",
+    "fresnel_zeros",
+    "fresnelc_zeros",
+    "fresnels_zeros",
+    "h1vp",
+    "h2vp",
+    "ivp",
+    "jn_zeros",
+    "jnjnp_zeros",
+    "jnp_zeros",
+    "jnyn_zeros",
+    "jvp",
+    "kei_zeros",
+    "keip_zeros",
+    "kelvin_zeros",
+    "ker_zeros",
+    "kerp_zeros",
+    "kvp",
+    "lmbda",
+    "lpmn",
+    "lpn",
+    "lqmn",
+    "lqn",
+    "mathieu_even_coef",
+    "mathieu_odd_coef",
+    "obl_cv_seq",
+    "pbdn_seq",
+    "pbdv_seq",
+    "pbvv_seq",
+    "perm",
+    "polygamma",
+    "pro_cv_seq",
+    "riccati_jn",
+    "riccati_yn",
+    "sinc",
+    "stirling2",
+    "y0_zeros",
+    "y1_zeros",
+    "y1p_zeros",
+    "yn_zeros",
+    "ynp_zeros",
+    "yvp",
+    "zeta",
+]
+
+_FACTORIALK_LIMITS_64BITS: dict[int, int]
+_FACTORIALK_LIMITS_32BITS: dict[int, int]
+
+digamma = psi
+
+def diric(x: NDArray, n: DecimalNumber) -> NDArray: ...
+def jnjnp_zeros(nt: int) -> _ZerosOrder: ...
+def jnyn_zeros(n: int, nt: int) -> _ZerosOrder: ...
+def jn_zeros(n: int, nt: int) -> NDArray: ...
+def jnp_zeros(n: int, nt: int) -> NDArray: ...
+def yn_zeros(n: int, nt: int) -> NDArray: ...
+def ynp_zeros(n: int, nt: int) -> NDArray: ...
+def y0_zeros(nt: int, complex: Literal[False] = ...) -> _PointsDerivate: ...
+def y1_zeros(nt: int, complex: Literal[False] = ...) -> _PointsDerivate: ...
+def y1p_zeros(nt: int, complex: Literal[False] = ...) -> _PointsDerivate: ...
+@overload
+def jvp(v: NDArray, z: _ComplexNumber | NDArray, n: Literal[1] = ...) -> NDArray: ...
+@overload
+def jvp(v: NDArray, z: _ComplexNumber | NDArray, n: int) -> NDArray: ...
+@overload
+def jvp(v: DecimalNumber | NDArray, z: NDArray, n: Literal[1] = ...) -> NDArray: ...
```

Comment:
These annotations should probably all be `ArrayLike` (from `numpy.typing`), rather than `DecimalNumber | NDArray`. See for example `special/_orthogonal.pyi`.

Basically if the input is annotated in the docstring as `array_like` and the implementation matches (i.e., `np.asarray` gets called on it), that is the annotation to use.